test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/ReflectionTest.scala

```diff
@@ -131,6 +133,103 @@ class ReflectionTest {
     if (hiddenJSObj.getClass() != null)
       fail("optimizer thought that hiddenJSObj.getClass() was non-null")
   }
+
+  @Test def jsTypesKeptOnlyForTheirData_Issue4850(): Unit = {
+    import JSTypesKeptOnlyForTheirData._
+
+    @noinline
+    def nameOf(cls: Class[_]): String = cls.getName()
+
+    @inline
+    def testName(expectedShortName: String, cls: Class[_]): Unit = {
+      val prefix = "org.scalajs.testsuite.compiler.ReflectionTest$JSTypesKeptOnlyForTheirData$"
+      val expectedName = prefix + expectedShortName
+
+      assertEquals(expectedName, cls.getName()) // constant-folded
+      assertEquals(expectedName, nameOf(cls)) // evaluated at run-time
+    }
+
+    testName("NativeClass", classOf[NativeClass])
+    testName("NativeObject$", classOf[Array[NativeObject.type]].getComponentType())
+    testName("NativeTrait", classOf[NativeTrait])
+    testName("NonNativeClass", classOf[NonNativeClass])
+    testName("NonNativeObject$", classOf[Array[NonNativeObject.type]].getComponentType())
+    testName("NonNativeTrait", classOf[NonNativeTrait])
+
+    @noinline
+    def isInterfaceOf(cls: Class[_]): Boolean = cls.isInterface()
+
+    @inline
+    def testIsInterface(expected: Boolean, cls: Class[_]): Unit = {
+      assertEquals(expected, cls.isInterface()) // could be constant-folded in the future
+      assertEquals(expected, isInterfaceOf(cls)) // evaluated at run-time
+    }
+
+    // Consistent with isInterfaceForInstantiatedJSTypes()
+    testIsInterface(false, classOf[NativeClass])
+    testIsInterface(false, classOf[Array[NativeObject.type]].getComponentType())
+    testIsInterface(false, classOf[NativeTrait])
+    testIsInterface(false, classOf[NonNativeClass])
+    testIsInterface(false, classOf[Array[NonNativeObject.type]].getComponentType())
+    testIsInterface(false, classOf[NonNativeTrait])
+
+    @noinline
+    def isInstance(cls: Class[_], x: Any): Boolean = cls.isInstance(x)
+
+    @inline
+    def testIsInstance(expected: Boolean, cls: Class[_], x: Any): Unit = {
+      assertEquals(expected, cls.isInstance(x))
+      assertEquals(expected, isInstance(cls, x))
+    }
+
+    @inline
+    def testIsInstanceThrows(expected: js.Dynamic, cls: Class[_], x: Any): Unit = {
+      val e1 = assertThrows(classOf[js.JavaScriptException], cls.isInstance(x))
+      assertTrue(e1.toString(), js.special.instanceof(e1.exception, expected))
+
+      val e2 = assertThrows(classOf[js.JavaScriptException], isInstance(cls, x))
+      assertTrue(e2.toString(), js.special.instanceof(e2.exception, expected))
+    }
+
+    val jsDate: Any = new js.Date(1684246473882.0)
+
+    testIsInstance(false, classOf[NonNativeClass], jsDate)
+    testIsInstance(true, classOf[JSDateForIsInstance], jsDate)
+
+    testIsInstanceThrows(js.constructorOf[js.ReferenceError], classOf[NativeClass], jsDate)
```

Comment:
To check my understanding: This throws ReferenceError, because `NativeClass` isn't actually defined, right?

Reply:
Yes, indeed. I added a comment.